broker/session.py

```diff
@@ -77,24 +77,30 @@ def shell(self, pty=False):
         channel = self.session.open_session()
         return InteractiveShell(channel, pty)
 
-    def sftp_read(self, source, destination=None):
+    def sftp_read(self, source, destination=None, return_data=False):
         """read a remote file into a local destination"""
-        if not destination:
-            destination = source
-        elif destination.endswith("/"):
-            destination = destination + Path(source).name
-        # create the destination path if it doesn't exist
-        destination = Path(destination)
-        destination.parent.mkdir(parents=True, exist_ok=True)
-        destination.touch()
+        if not return_data:
+            if not destination:
+                destination = source
+            elif destination.endswith("/"):
+                destination = destination + Path(source).name
+            # create the destination path if it doesn't exist
+            destination = Path(destination)
+            destination.parent.mkdir(parents=True, exist_ok=True)
         # initiate the sftp session, read data, write it to a local destination
         sftp = self.session.sftp_init()
         with sftp.open(
             source, ssh2_sftp.LIBSSH2_FXF_READ, ssh2_sftp.LIBSSH2_SFTP_S_IRUSR
         ) as remote:
-            with destination.open("wb") as local:
-                for size, data in remote:
-                    local.write(data)
+            if return_data:
+                captured_data = bytes()
+                for rc, data in remote:
+                    captured_data += data
+                return captured_data
+            else:
+                with destination.open("wb") as local:
+                    for rc, data in remote:
+                        local.write(data)
```

Comment:
We likely can simplify this a bit more. Since we are returning as bytes, we can read it into the object then either return that object or write it to the destination file.
```suggestion
            captured_data = bytes()
            for rc, data in remote:
                captured_data += data
            if return_data:
                return captured_data
            destination.write_bytes(data)
```